### PR TITLE
Add TSOA, generate spec and client code

### LINKED ISSUE #57: create dockerized processes to generate server and client stubs
openApi publishes an image of their [generator CLI](https://github.com/OpenAPITools/openapi-generator#16---docker) which we have used successfully in Hedwig to generate typescript fetch client code. It also has option to generate `graphql-nodejs-express-server` and `nodejs-express-server (beta)` servers. 

there is also a [swagger+node specific codegen package](https://www.npmjs.com/package/swagger-node-codegen) that can be used as a CLI tool

Either way, we will need to generate types for the generated server code as there's not currently support for typescript server generation. Maybe this: https://github.com/horiuchi/dtsgenerator

NOTE: If generating typescript server is funky, we can always switch to server code first -> generated spec -> generated client stubs with https://www.npmjs.com/package/tsoa

**AC**
- Add service(s) to docker compose project that can
  - easily, reproducibly, repeatedly generate client code
  - easily, reproducibly, repeatedly generate server stubs

### PR DESCRIPTION
**Pull request by cailyncodes**
Should close #57 

Exciting times!

### CODE REVIEW

File: src/controllers/UserController.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  Get,
+	Path,
+	Request,
+  Route,
+} from "tsoa";
+import { User } from "../models/User";
+import { UsersService } from "../services/user/UserService";
+
+@Route("users")
+export class UsersController extends Controller {
```
Comment (jlhogan) [inline src/controllers/UserController.ts:12]: so tsoa generates all routes, and hooks up controllers to routes based on these decorators?

File: src/controllers/UserController.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+	@Get("current")
+	public async getCurrentUser(
+		@Request() req: Express.Request
+	): Promise<User> {
```
Comment (jlhogan) [inline src/controllers/UserController.ts:17]: and presumably tsoa decides which models to expose in the spec based on what is returned by the linked up controller functions

File: src/controllers/UserController.ts (hunk 1 above)
Comment (cailyncodes) [reply src/controllers/UserController.ts:12]: Yee

File: src/controllers/UserController.ts (hunk 2 above)
Comment (cailyncodes) [reply src/controllers/UserController.ts:17]: Tis my guess

File: src/services/user/UserService.ts [hunk 3]
```diff
@@ -0,0 +1,8 @@
+import { User } from "../../models/User";
+import { users } from "../../data/users";
+
+export class UsersService {
```
Comment (JordanGuinn) [inline src/services/user/UserService.ts:4]: extreme nit: class name `UsersService` vs file name `UserService.ts`

File: src/controllers/UserController.ts (hunk 1 above)
Comment (JordanGuinn) [inline src/controllers/UserController.ts:12]: extreme nit #2: file name `UserController.ts` vs class name `UsersController`

File: src/controllers/UserController.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+		const wingedKeysId = req.sub;
```
Comment (JordanGuinn) [inline src/controllers/UserController.ts:18]: Out of curiosity, is there an annotation that can be used to mark this particular request param?

File: src/controllers/UserController.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+	Request,
+  Route,
+} from "tsoa";
+import { User } from "../models/User";
+import { UsersService } from "../services/user/UserService";
+
+@Route("users")
+export class UsersController extends Controller {
+
+	@Get("current")
```
Comment (JordanGuinn) [inline src/controllers/UserController.ts:14]: Love how intuitive these annotations are 👏🏽👏🏽

File: src/services/user/UserService.ts (hunk 3 above)
Comment (cailyncodes) [reply src/services/user/UserService.ts:4]: Ooh thanks! Good catch

File: src/controllers/UserController.ts (hunk 4 above)
Comment (cailyncodes) [reply src/controllers/UserController.ts:18]: I don't think I follow this question?

Comment (JordanGuinn) [reply src/controllers/UserController.ts:18]: I'm not sure where `sub` comes from or how it's supposed to be provided by the client here, so I'm wondering if that needs to captured in our swagger spec at all?  And assuming it does, I'm wondering if there are any annotations we can use (similar to `@Request` or `@Route`) to make that happen?

Comment (cailyncodes) [reply src/controllers/UserController.ts]: Check out the `custom.d.ts` file -- we are extending the included `Express.Request` type. (After integrating with OpenAPI auth middleware, this has become `req.User` with our defined `User` object (also documented in `custom.d.ts`)

Comment (JordanGuinn) [reply src/controllers/UserController.ts]: Okay I think I see now - the user is retrieved with the ID made available after decoding the jwt, and is then subsequently attached to the request as part of our middleware run.  Def no need to document that! 🙌🏽🙌🏽

File: src/index.ts [hunk 6]
```diff
@@ -16,7 +18,14 @@ const app = express();
 app.use(json());
 
 /* Register the API middleware */
-app.use('/api', api);
+// Require authentication for all API requests
```
Comment (jlhogan) [inline src/index.ts:21]: nit clean up commented out code

File: src/router/api.ts [hunk 7]
```diff
@@ -2,22 +2,17 @@ import express from 'express';
 import multer from 'multer';
 import path from 'path';
 import { handleError } from '../middleware/error';
-import { authenticate } from "../middleware/authenticate";
-import { getUserByWingedKeysId } from '../data/users';
+// import { authenticate } from "../middleware/authenticate";
```
Comment (jlhogan) [inline src/router/api.ts:5]: nit clean up

File: src/router/api.ts [hunk 8]
```diff
@@ -2,22 +2,17 @@ import express from 'express';
...
+
+// TODO: Combine file upload with TSOA
```
Comment (jlhogan) [inline src/router/api.ts:7]: meaning  create decorated controller for file upload route?

Comment (cailyncodes) [reply src/router/api.ts:7]: This todo is for another ticket -- this entire file would be scrapped and file upload needs to be added to a controller that is actually used

Comment (jlhogan) [reply src/router/api.ts:7]: like.... this ticket https://github.com/ctoec/data-collection/issues/73 :)

Comment (cailyncodes) [reply src/router/api.ts:7]: Oops lol realizing I did not really answer your question but yes 73 is great

**Pull request merged**